iceoryx2-ffi/cxx/include/iox2/sample_mut.hpp

```diff
@@ -13,120 +13,220 @@
     auto operator=(const SampleMut&) -> SampleMut& = delete;
 
-    auto header() const -> const HeaderPublishSubscribe& {
-        IOX_TODO();
-    }
-    auto user_header() const -> const UserHeader& {
-        IOX_TODO();
-    }
-    auto user_header_mut() -> UserHeader& {
-        IOX_TODO();
-    }
-    auto payload() const -> const Payload& {
-        IOX_TODO();
-    }
-    auto payload_mut() -> Payload& {
-        IOX_TODO();
-    }
-    void write_payload(const Payload& payload) {
-        IOX_TODO();
-    }
-};
+    /// Returns a const reference to the payload of the [`Sample`]
+    auto operator*() const -> const Payload&;
 
-template <ServiceType S, typename Payload>
-class SampleMut<S, Payload, void> {
-  public:
-    SampleMut() = default;
-    SampleMut(SampleMut&&) = default;
-    auto operator=(SampleMut&&) -> SampleMut& = default;
-    ~SampleMut() = default;
+    /// Returns a reference to the payload of the [`Sample`]
+    auto operator*() -> Payload&;
 
-    SampleMut(const SampleMut&) = delete;
-    auto operator=(const SampleMut&) -> SampleMut& = delete;
+    /// Returns a const pointer to the payload of the [`Sample`]
+    auto operator->() const -> const Payload*;
 
-    auto header() const -> const HeaderPublishSubscribe& {
-        IOX_TODO();
-    }
-    auto payload() const -> const Payload& {
-        IOX_TODO();
-    }
-    auto payload_mut() -> Payload& {
-        IOX_TODO();
-    }
-    void write_payload(const Payload& payload) {
-        IOX_TODO();
-    }
-};
+    /// Returns a pointer to the payload of the [`Sample`]
+    auto operator->() -> Payload*;
 
-template <ServiceType S, typename Payload>
-class SampleMut<S, iox::Slice<Payload>, void> {
-  public:
-    SampleMut() = default;
-    SampleMut(SampleMut&&) = default;
-    auto operator=(SampleMut&&) -> SampleMut& = default;
-    ~SampleMut() = default;
+    /// Returns a reference to the [`Header`] of the [`Sample`].
+    auto header() const -> const HeaderPublishSubscribe&;
 
-    SampleMut(const SampleMut&) = delete;
-    auto operator=(const SampleMut&) -> SampleMut& = delete;
+    /// Returns a reference to the user_header of the [`Sample`]
+    template <typename T = UserHeader, typename = std::enable_if_t<!std::is_same_v<void, UserHeader>, T>>
+    auto user_header() const -> const T&;
 
-    auto header() const -> const HeaderPublishSubscribe& {
-        IOX_TODO();
-    }
-    auto payload() const -> const Payload& {
-        IOX_TODO();
-    }
-    auto payload_mut() -> Payload& {
-        IOX_TODO();
+    /// Returns a mutable reference to the user_header of the [`Sample`].
+    template <typename T = UserHeader, typename = std::enable_if_t<!std::is_same_v<void, UserHeader>, T>>
+    auto user_header_mut() -> T&;
+
+    /// Returns a reference to the const payload of the sample.
+    auto payload() const -> const Payload&;
+
+    /// Returns a reference to the payload of the sample.
+    auto payload_mut() -> Payload&;
```

Comment:
Do we want to introduce this terminology in C++.

I'm not suggestion to replace it, I'm just curious if it would make sense to stick to C++ idioms/terminology?

But I must confess that I'm lightly leaning towards the terminology in this PR.

Reply:
I am introducing this terminology to be as consistent as possible to the Rust API. In this sense it is much easier to switch between C++ and Rust (for instance when code must be ported).
Also the Rust documentation complements the C++ docu nicely when the APIs are closely together.